worker/src/main/java/gov/cms/ab2d/worker/service/JobProcessingServiceImpl.java

```diff
@@ -128,19 +139,52 @@ public Job processJob(final String jobId) {
         // A mutex lock that all threads for a contract uses while writing into the shared files
         var lock = new ReentrantLock();
 
-        final List<Future<Integer>> futureResourcesHandles = patients.stream()
-                .map(patient -> patient.getPatientId())
-                .filter(patientId -> !isOptOutPatient(patientId))
-                .map(patientId -> patientClaimsProcessor.process(patientId, lock, outputFile, errorFile))
-                .collect(Collectors.toList());
+        int errorCount = 0;
+        JobStatus jobStatus = null;
+
+        int recordsProcessedCount = 0;
+        final List<Future<Integer>> futureResourcesHandles = new ArrayList<>();
+        for (GetPatientsByContractResponse.PatientDTO patient : patients) {
+            ++recordsProcessedCount;
+
+            final String patientId = patient.getPatientId();
+
+            if (isOptOutPatient(patientId)) {
+                // this patient has opted out. skip patient record.
+                continue;
+            }
+
+            futureResourcesHandles.add(patientClaimsProcessor.process(patientId, lock, outputFile, errorFile));
+
+            if (recordsProcessedCount % cancellationCheckFrequency == 0) {
+                errorCount += processHandles(futureResourcesHandles);
+
+                // A Job could run for a long time perhaps hours.
+                // While the job is in progress, the job could be cancelled.
+                // So the worker needs to periodically check the job status to ensure it has not been cancelled.
+
+                jobStatus = jobRepository.findJobStatus(jobUuid);
+                if (jobHasBeenCancelled(jobStatus)) {
+                    log.warn("Job [{}] has been cancelled. Attempting to stop processing the job shortly ... ", jobUuid);
+                    break;
+                }
+
+            }
+        }
 
 
-        int errorCount = processHandles(futureResourcesHandles);
         while (!futureResourcesHandles.isEmpty()) {
             sleep();
             errorCount += processHandles(futureResourcesHandles);
         }
 
+        if (jobHasBeenCancelled(jobStatus)) {
+            final String errMsg = "Job was cancelled while it was being processed";
+            log.warn("{} - JobUuid :[{}]", errMsg, jobUuid);
+            throw new JobCancelledException(errMsg);
```

Comment:
An exception in this method will not really stop all the Futures from being processed by the Thread Pool. They will continue to be executed in the background.